File: ios/RNModules/RNDepthCamera/DepthCameraViewManager.m
#import <React/RCTBridge.h>
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_REMAP_MODULE(DepthCameraView, DepthCameraViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(detectionWidthScale, float);
RCT_EXPORT_VIEW_PROPERTY(colorMode, int);
RCT_EXPORT_VIEW_PROPERTY(minDistanceDetection, BOOL);
RCT_EXPORT_VIEW_PROPERTY(smoothed, BOOL);
RCT_EXPORT_VIEW_PROPERTY(enabled, BOOL);

RCT_EXPORT_VIEW_PROPERTY(onReady, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onPause, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onMinDistance, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onCameraSize, RCTDirectEventBlock);

RCT_EXTERN_METHOD(supports:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(takePicture:(nonnull NSNumber *)node
                  options:(NSDictionary *)options
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/RNModules/RNDepthCamera/DepthCameraViewManager.swift
import React
import Foundation
import ARKit

@objc(DepthCameraViewManager)
class DepthCameraViewManager: RCTViewManager {
    override static func requiresMainQueueSetup() -> Bool {
      return true
    }

    override func view() -> UIView {
        return DepthCameraView()
    }

    override var methodQueue: DispatchQueue! {
      return DispatchQueue.main
    }
  
    @objc(takePicture:reactTag:resolver:rejecter:)
    func takePicture(options: NSDictionary, reactTag: NSNumber, resolve: @escaping RCTPromiseResolveBlock,reject: @escaping RCTPromiseRejectBlock) {
      bridge.uiManager.prependUIBlock({_ , viewRegistry in
          let view = viewRegistry?[reactTag]
          if !(view is DepthCameraView) {
            reject("ERROR_TAKE", "Invalid view returned from registry", nil)
          } else if let view = view as? DepthCameraView {
            view.takePicture(options: options, resolve: resolve, reject: reject)
          }
      })
    }

    @objc
    func supports(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth) {
          resolve(true)
        } else {
          resolve(false)
        }
    }
}
